src/toil/provisioners/clusterScaler.py

```diff
@@ -302,36 +298,66 @@ def __init__(self, scaler, preemptable):
 
-        if isinstance(self.scaler.jobBatcher.batchSystem, AbstractScalableBatchSystem):
-            totalNodes = len(self.scaler.jobBatcher.batchSystem.getNodes(self.preemptable))
-        else:
-            totalNodes = 0
-        logger.info('Starting with %s node(s) in the cluster.', totalNodes)
         while not self.scaler.stop:
             with throttle(self.scaler.config.scaleInterval):
-                # Calculate the approx. number nodes needed
-                # TODO: Correct for jobs already running which can be considered fractions of a job
-                queueSize = self.scaler.jobBatcher.getNumberOfJobsIssued(preemptable=self.preemptable)
-
+                # Estimate the number of nodes to run the issued jobs.
+            
+                # Number of jobs issued
+                queueSize = self.scaler.leader.getNumberOfJobsIssued(preemptable=self.preemptable)
+                
+                # Job shapes of completed jobs
                 recentJobShapes = self.jobShapes.get()
                 assert len(recentJobShapes) > 0
+                
+                # Estimate of number of nodes needed to run recent jobs
                 nodesToRunRecentJobs = binPacking(recentJobShapes, self.nodeShape)
+                
+                # Actual calculation of the estimated number of nodes required
                 estimatedNodes = 0 if queueSize == 0 else max(1, int(round(
                     self.scaler.config.alphaPacking
                     * nodesToRunRecentJobs
-                    * float(queueSize)
-                    / len(recentJobShapes))))
+                    * float(queueSize) / len(recentJobShapes))))
+                
+                # Account for case where the average historical runtime of completed jobs is less
+                # than the runtime of currently running jobs. This is important
+                # to avoid a deadlock where the estimated number of nodes to run the jobs
+                # is too small to schedule a set service jobs and their dependent jobs, leading
+                # to service jobs running indefinitely.
+                
+                # How many jobs are currently running and their average runtime.
+                numberOfRunningJobs, currentAvgRuntime  = self.scaler.leader.getNumberAndAvgRuntimeOfCurrentlyRunningJobs()
+                
+                # Average runtime of recently completed jobs
+                historicalAvgRuntime = sum(map(lambda jS : jS.wallTime, recentJobShapes))
+            
+                # Ratio of avg. runtime of currently running and completed jobs
+                runtimeCorrection = float(currentAvgRuntime)/historicalAvgRuntime if currentAvgRuntime > historicalAvgRuntime and numberOfRunningJobs >= estimatedNodes else 1.0
+                
+                # Make correction, if necessary (only do so if cluster is busy and average runtime is higher than historical
+                # average)
+                if runtimeCorrection != 1.0:
```

Comment:
This will be triggered a bunch, but I still think it's a better way to adjust our prior on how long we expect the cluster to run. Remember when we first start up the cluster it can take ages before it has a decent set of jobs to properly estimate the runtime of queued jobs. This heuristic will tend to increase the cluster size to tackle this case.